tests/e2e/test_runtime.py

```diff
@@ -102,3 +103,114 @@ def test_disabled_rules():
         assert (
             "Disabled rule should not have fired" not in result.stdout
         ), "A disabled rule fired unexpectedly"
+
+
+@pytest.mark.e2e
+@pytest.mark.parametrize(
+    "shutdown_delay",
+    [
+        pytest.param(
+            None,
+            id="default_timeout",
+        ),
+        pytest.param(
+            5,
+            id="custom_timeout",
+        ),
+    ],
+)
+def test_terminate_process_source_end(update_environment, shutdown_delay):
+    """
+    Execute a rulebook and ensure ansible-rulebook terminates correctly
+    after an event source plugin ends
+    """
+
+    if shutdown_delay:
+        env = update_environment({"EDA_SHUTDOWN_DELAY": str(shutdown_delay)})
+
+    rulebook = utils.BASE_DATA_PATH / "rulebooks/test_process_source_end.yml"
+    cmd = utils.Command(rulebook=rulebook)
+
+    LOGGER.info(f"Running command: {cmd}")
+    result = subprocess.run(
+        cmd,
+        timeout=DEFAULT_CMD_TIMEOUT,
+        capture_output=True,
+        cwd=utils.BASE_DATA_PATH,
+        text=True,
+        env=env if shutdown_delay else None,
+    )
+
+    assert result.returncode == 0
+    assert not result.stderr
+
+    with check:
+        assert (
+            len(
+                [
+                    line
+                    for line in result.stdout.splitlines()
+                    if '"msg": "Sleeping..."' in line
+                ]
+            )
+            == 1
+        ), "long-running playbook failed to fire"
+
+    with check:
+        assert (
+            len(
+                [
+                    line
+                    for line in result.stdout.splitlines()
+                    if "Parallel action triggered successfully" in line
+                ]
+            )
+            == 3
+        ), "a parallel action failed to fire"
+
+    with check:
+        assert (
+            len(
+                [
+                    line
+                    for line in result.stdout.splitlines()
+                    if '"msg": "Rise and shine..."' in line
+                ]
+            )
+            == 1
+        ), "long-running playbook failed to finish"
+
+
+@pytest.mark.e2e
+def test_terminate_process_sigint():
+    """
+    Execute a rulebook and ensure ansible-rulebook terminates correctly
+    after a SIGINT is issued
+    """
+
+    rulebook = utils.BASE_DATA_PATH / "rulebooks/test_process_sigint.yml"
+    cmd = utils.Command(rulebook=rulebook)
+
+    LOGGER.info(f"Running command: {cmd}")
+
+    process = subprocess.Popen(
+        cmd,
+        cwd=utils.BASE_DATA_PATH,
+        text=True,
+        stdout=subprocess.PIPE,
+    )
+
+    # loop for 10 iterations until we get output, then SIGINT
+    i = 0
+    while i < 10:
+        line = process.stdout.readline()
+        if "{'action': 'long_loop'}" in line:
+            process.send_signal(subprocess.signal.SIGINT)
+            process.wait()
+            break
+        i += 1
+        sleep(0.5)
+    else:
+        process.kill()
```

Comment:
I don't understand this loop, it adds an arbitrary timeout outside the scope of the conf. Why are you not using the DEFAULT_CMD_TIMEOUT?

```
    start = time.time()
    while line := process.stdout.readline():
        if "{'action': 'long_loop'}" in line:
            break
        time.sleep(0.1)
        if time.time() - start > DEFAULT_CMD_TIMEOUT:
            process.kill()
```

Reply:
It's a good idea, I'll use that :+1: